cmd/gen-go-sample/inittree.go

```diff
@@ -106,6 +109,29 @@ func (t *initTree) get(k string, info pbinfo.Info) (*initTree, error) {
 		} else {
 			// type is a message
 			v.typ.desc = typ
+
+			if d, ok := typ.(*descriptor.DescriptorProto); ok {
+				if d.GetOptions().GetMapEntry() == true {
+					// type is a map entry
+					for _, f2 := range d.Field {
+						if f2.GetName() == "key" {
+							// key is always a primitive type
+							v.typ.keyType = &initType{prim: f2.GetType()}
+						}
+						if f2.GetName() == "value" {
+							if tn2 := f.GetTypeName(); tn2 == "" {
```

Comment:
Do you mean `f2.GetTypeName()` here? `f` is the outer Field variable.

Nit: Have you considered using [GetType](https://godoc.org/github.com/golang/protobuf/protoc-gen-go/descriptor#FieldDescriptorProto.GetType) & comparing to the type const? Personally, I think it would be easier to grok this code. (I know that TypeName is only set for Messages & Enums, but that requires looking at the FieldDescriptorProto.type_name comments)

```go
v.typ.valueType = &initType{prim: f2.GetType()}

if f2.GetType() == descriptor.FieldDescriptorProto_TYPE_MESSAGE {
  tn2 := f2.GetTypeName()
  typ := info.Type[tn2]

  if typ == nil {
    return nil, errors.E(nil, "cannot find descriptor of %q", tn2)
  }

  v.typ.valueType = &initType{desc: typ}
}
```

Reply:
Done.